Use java.util.Base64 for SSL cert / key encoding when running on Java8+

normanmaurer opened the pull request:
Motivation:

We need to encode certs and keys when creating *OpenSsl*Context instances. This is kind of "slow" when using a high leak-detector level and so slow down testing a lot where a lot of these are created and destroyed. We can speed this up by using the java.util.Base64.Encoder when we run on Java8+.

Modifications:

Use java.util.Base64.Encoder when possible.

Result:

Much faster test times when using Java8+. Fixes [#7029].

Before (when running tests in handler module):

[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 20:23 min
[INFO] Finished at: 2017-07-28T09:26:57+02:00
[INFO] Final Memory: 48M/1301M
[INFO] ---

After (when running tests in handler module):

[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 05:09 min
[INFO] Finished at: 2017-07-28T08:45:14+02:00
[INFO] Final Memory: 38M/820M
[INFO] ------------------------------------------------------------------------

normanmaurer commented:
This will also greatly improve build times on the ci for PRs 👍

johnou commented:
Impressive speed up, nice nagging @doom369 ;)

normanmaurer commented:
Need to investigate the build errors... pass locally :(

ninja- commented:
@normanmaurer the build fails with resource leak detected...probably unrelated to this pr :v

Scottmitch commented:
@normanmaurer - What is the performance difference between the JDK's encoder and Netty's encoder without leak detection? This PR will unconditionally uses the JDK's encoder on JDK8+ even in production even though this PR is only designed to work around the high overhead of high levels of resource leak detection. This would mean we may loose the claimed benefits of the pooled allocator in production. I wonder:

- If Netty's encoder is slower than the JDK's in general ... can we improve it?
- Instead of using the JDK's encoder as a work around  can we punch a hole in the leak detection here? This is effectively what we are doing by using the JDK's algorithm ... but we lose the potential benefits of the allocator in the process.

doom369 commented:
@Scottmitch just made quick benchmark:

```
@State(Scope.Benchmark)
@Warmup(iterations = 10)
@Measurement(iterations = 20)
public class SslUtilBenchmark extends AbstractMicrobenchmark {

    static {
        ResourceLeakDetector.setLevel(ResourceLeakDetector.Level.DISABLED);
    }

    private ByteBuf src;


    @Setup
    public void setup() {
        src = Unpooled.wrappedBuffer("Hello Worlds".getBytes(CharsetUtil.US_ASCII));
    }

    @TearDown
    public void close() {
        src.release();
    }

    @Benchmark
    public ByteBuf java8ToBase64() {
        return toBase64Java8(src);
    }

    @Benchmark
    public ByteBuf oldToBase64() {
        return oldToBase(PooledByteBufAllocator.DEFAULT, src);
    }

    private static ByteBuf oldToBase(ByteBufAllocator allocator, ByteBuf src) {
        ByteBuf dst = Base64.encode(src, src.readerIndex(),
                src.readableBytes(), true, Base64Dialect.STANDARD, allocator);
        src.readerIndex(src.writerIndex());
        return dst;
    }

    private static ByteBuf toBase64Java8(ByteBuf src) {
        ByteBuffer srcBuffer = src.internalNioBuffer(src.readerIndex(), src.readableBytes());
        int remaining = srcBuffer.remaining();
        ByteBuffer encodedBuffer = java.util.Base64.getEncoder().encode(srcBuffer);
        src.skipBytes(remaining - srcBuffer.remaining());
        return Unpooled.wrappedBuffer(encodedBuffer);
    }

}
```

Result:

```
Benchmark                        Mode  Cnt         Score        Error  Units
SslUtilBenchmark.java8ToBase64  thrpt   40  31522693.400 ± 779193.722  ops/s
SslUtilBenchmark.oldToBase64    thrpt   40  11087806.505 ± 343671.742  ops/s
```

johnou commented:
@doom369 wouldn't it be more interesting with the leak detection set as paranoid

doom369 commented:
@johnou I think main concern of @Scottmitch was regarding real apps that doesn't use paranoid in production. However, here is result for above test in prarnoid mode:

```
Benchmark                        Mode  Cnt         Score         Error  Units
SslUtilBenchmark.java8ToBase64  thrpt   40  28628813.742 ± 1118104.512  ops/s
SslUtilBenchmark.oldToBase64    thrpt   40     10036.239 ±     678.807  ops/s
```

Scottmitch commented:
> @johnou I think main concern of @Scottmitch was regarding real apps that doesn't use paranoid in production

+1

> ByteBuffer srcBuffer = src.internalNioBuffer(src.readerIndex(), src.readableBytes());

The code is currently not using `internalNioBuffer`but instead `nioBuffer`.

Requests for enhancements to the benchmark:
- Vary the size of input. Netty's encode/decode uses methods like `setInt` instead of always doing byte operations. As input size gets larger the impact may be more apparent.
- Try direct buffers. Netty's encode/decode can avoid copy operations for direct buffers, but JDK's always copies and requires heap buffers. We also ultimately want direct memory for the PEM encoded key because we will use it in JNI, so direct buffers can avoid additional copies.
- Encoded result length vary from estimate length. Both implementations estimate the total length of the buffer, but may over estimate. JDK's implementation makes a copy of the buffer and Netty just takes a slice.

Scottmitch commented:
@doom369 - You should also release the ByteBufs in the benchmarks.

normanmaurer commented:
ok fixed the build..

@Scottmitch I would not be concerned about switching to the Java provided Base64 encoder as it will only be used for encoding keys / certs and so will not happen frequently.  The problem with the one in netty is that it does a lot of ByteBuf operations and so is slow with paranoid leak detection

netkins commented on handler/src/main/java/io/netty/handler/ssl/Java8SslUtils.java:
```diff
@@ -15,10 +15,17 @@
 package io.netty.handler.ssl;
 
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'io.netty.buffer.ByteBufAllocator'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

netkins replied:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'io.netty.buffer.ByteBufAllocator'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

netkins commented on handler/src/main/java/io/netty/handler/ssl/Java8SslUtils.java:
```diff
@@ -15,10 +15,17 @@
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.buffer.Unpooled;
+import io.netty.handler.codec.base64.Base64;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'io.netty.handler.codec.base64.Base64'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

netkins replied:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'io.netty.handler.codec.base64.Base64'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

netkins commented on handler/src/main/java/io/netty/handler/ssl/Java8SslUtils.java:
```diff
@@ -15,10 +15,17 @@
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.buffer.Unpooled;
+import io.netty.handler.codec.base64.Base64;
+import io.netty.handler.codec.base64.Base64Dialect;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'io.netty.handler.codec.base64.Base64Dialect'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

netkins replied:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'io.netty.handler.codec.base64.Base64Dialect'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

Scottmitch commented:
> I would not be concerned about switching to the Java provided Base64 encoder

Perhaps but it would make me feel much better to understand what we are losing/gaining (see https://github.com/netty/netty/pull/7030#issuecomment-318842190)).

> as it will only be used for encoding keys / certs and so will not happen frequently

It depends upon the use case and it maybe more common than we think (/cc @rkapsi @slandelle)?

rkapsi commented:
Hey - thx for the heads up @Scottmitch 

In our case it's definitively very frequently (many millions of times per day). Not sure how much of an impact it'd have in terms of memory usage/GC pressure.

slandelle commented:
> It depends upon the use case and it maybe more common than we think (/cc @rkapsi @slandelle)?

In Gatling, I do have several use cases for spawning tons of SslContext.

Also, I'm being curious as to how `java.util.Base64` encoding is that faster than Netty implementation, all the more as Netty should be able to save an array copy.

fenik17 commented:
> I'm being curious as to how java.util.Base64 encoding is that faster than Netty implementation

Netty's implementation do a many virtual methods calls, within which there are checks of ranges and the buffers availability. It's not free...

Scottmitch commented:
> I'm being curious as to how java.util.Base64 encoding is that faster than Netty implementation

@slandelle - Please look at my comments for benchmark modifications https://github.com/netty/netty/pull/7030#issuecomment-318851991 and https://github.com/netty/netty/pull/7030#issuecomment-318853787. I think the benchmark posted in https://github.com/netty/netty/pull/7030#issuecomment-318845138 by @doom369 needs some work ... @normanmaurer can you incorporate the benchmark into this PR and make the recommended changes?

normanmaurer commented:
Yep will do

> Am 01.08.2017 um 18:34 schrieb Scott Mitchell <notifications@github.com>:
> 
> I'm being curious as to how java.util.Base64 encoding is that faster than Netty implementation
> 
> @slandelle - Please look at my comments for benchmark modifications #7030 (comment) and #7030 (comment). I think the benchmark posted in #7030 (comment) by @doom369 needs some work ... @normanmaurer can you incorporate the benchmark into this PR and make the recommended changes?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

normanmaurer commented:
will do this tomorrow first... Sorry super busy atm

normanmaurer commented:
Here are there results:

```
Benchmark                                          (level)  (size)   Mode  Cnt         Score         Error  Units
Base64Benchmark.javaBase64Encode                  DISABLED    1024  thrpt   40  20800602.379 ?  163282.349  ops/s
Base64Benchmark.javaBase64Encode                  DISABLED   65536  thrpt   40  20803726.200 ?  217445.850  ops/s
Base64Benchmark.javaBase64Encode                    SIMPLE    1024  thrpt   40  25202718.406 ? 1762087.810  ops/s
Base64Benchmark.javaBase64Encode                    SIMPLE   65536  thrpt   40  22675633.133 ?  112909.173  ops/s
Base64Benchmark.javaBase64Encode                  ADVANCED    1024  thrpt   40  22572037.666 ?  130977.178  ops/s
Base64Benchmark.javaBase64Encode                  ADVANCED   65536  thrpt   40  27710505.933 ?  455142.496  ops/s
Base64Benchmark.javaBase64Encode                  PARANOID    1024  thrpt   40  25530062.008 ? 1648212.492  ops/s
Base64Benchmark.javaBase64Encode                  PARANOID   65536  thrpt   40  25360766.887 ? 1596951.221  ops/s
Base64Benchmark.nettyBase64EncodeDirectDst        DISABLED    1024  thrpt   40   7494990.030 ?   26093.074  ops/s
Base64Benchmark.nettyBase64EncodeDirectDst        DISABLED   65536  thrpt   40   7431155.791 ?   59203.640  ops/s
Base64Benchmark.nettyBase64EncodeDirectDst          SIMPLE    1024  thrpt   40   7219401.183 ?   73898.002  ops/s
Base64Benchmark.nettyBase64EncodeDirectDst          SIMPLE   65536  thrpt   40   7193255.869 ?   41972.278  ops/s
Base64Benchmark.nettyBase64EncodeDirectDst        ADVANCED    1024  thrpt   40   1239037.369 ?   15647.228  ops/s
Base64Benchmark.nettyBase64EncodeDirectDst        ADVANCED   65536  thrpt   40   1239422.881 ?   13413.463  ops/s
Base64Benchmark.nettyBase64EncodeDirectDst        PARANOID    1024  thrpt   40     11691.309 ?     152.899  ops/s
Base64Benchmark.nettyBase64EncodeDirectDst        PARANOID   65536  thrpt   40     11662.815 ?     152.659  ops/s
Base64Benchmark.nettyBase64EncodeHeapDst          DISABLED    1024  thrpt   40  18614819.078 ?  151177.357  ops/s
Base64Benchmark.nettyBase64EncodeHeapDst          DISABLED   65536  thrpt   40  18655986.503 ?   71473.504  ops/s
Base64Benchmark.nettyBase64EncodeHeapDst            SIMPLE    1024  thrpt   40  18659288.151 ?   74717.457  ops/s
Base64Benchmark.nettyBase64EncodeHeapDst            SIMPLE   65536  thrpt   40  18572729.925 ?  105130.834  ops/s
Base64Benchmark.nettyBase64EncodeHeapDst          ADVANCED    1024  thrpt   40  18466515.773 ?   65368.858  ops/s
Base64Benchmark.nettyBase64EncodeHeapDst          ADVANCED   65536  thrpt   40  18390339.120 ?   92829.343  ops/s
Base64Benchmark.nettyBase64EncodeHeapDst          PARANOID    1024  thrpt   40  17731510.487 ?  335241.762  ops/s
Base64Benchmark.nettyBase64EncodeHeapDst          PARANOID   65536  thrpt   40  17841683.741 ?  235045.737  ops/s
Base64Benchmark.nettyBase64EncodePooledDirectDst  DISABLED    1024  thrpt   40  12393784.383 ?  171198.152  ops/s
Base64Benchmark.nettyBase64EncodePooledDirectDst  DISABLED   65536  thrpt   40  12761150.326 ?   98524.888  ops/s
Base64Benchmark.nettyBase64EncodePooledDirectDst    SIMPLE    1024  thrpt   40  12652585.024 ?  121576.528  ops/s
Base64Benchmark.nettyBase64EncodePooledDirectDst    SIMPLE   65536  thrpt   40  12306853.857 ?   70824.722  ops/s
Base64Benchmark.nettyBase64EncodePooledDirectDst  ADVANCED    1024  thrpt   40   1321198.669 ?   15890.673  ops/s
Base64Benchmark.nettyBase64EncodePooledDirectDst  ADVANCED   65536  thrpt   40   1338972.365 ?   18762.021  ops/s
Base64Benchmark.nettyBase64EncodePooledDirectDst  PARANOID    1024  thrpt   40     11551.857 ?     125.186  ops/s
Base64Benchmark.nettyBase64EncodePooledDirectDst  PARANOID   65536  thrpt   40     11584.078 ?     149.223  ops/s

Benchmark result is saved to /Users/norman/Documents/workspace/netty/microbench/target/reports/performance/Base64Benchmark.json
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 1,958.119 sec - in io.netty.handler.codec.Base64Benchmark

Results :

Tests run: 1, Failures: 0, Errors: 0, Skipped: 0
```

johnou commented:
Maybe Oracle finally got something right ;)

normanmaurer commented:
one thing to note here is that with the JDK version its not possible to tell if it should use a direct or heap buffer for the output. Which means in situations where we want to write it to a Channel we will need to pay for a memory copy.

Scottmitch commented on microbench/src/main/java/io/netty/handler/codec/Base64Benchmark.java:
```diff
@@ -0,0 +1,108 @@
+
+        byte[] bytes = new byte[size];
+        ThreadLocalRandom.current().nextBytes(bytes);
+        src = Unpooled.wrappedBuffer(bytes);
```
we should test when this is a direct buffer too. The JDK will incur an additional copy because it only operates on heap data.

normanmaurer replied:
+1

Scottmitch commented on microbench/src/main/java/io/netty/handler/codec/Base64Benchmark.java:
```diff
@@ -0,0 +1,108 @@
+        int remaining = srcBuffer.remaining();
+        ByteBuffer encodedBuffer = java.util.Base64.getMimeEncoder(76, new byte[]{ '\n' }).encode(srcBuffer);
+        src.skipBytes(remaining - srcBuffer.remaining());
+        return Unpooled.wrappedBuffer(encodedBuffer);
```
can we also measure the impacts of going back to direct buffer ... the JDK will always produce heap based buffers and we need them to be direct.

Scottmitch commented on microbench/src/main/java/io/netty/handler/codec/Base64Benchmark.java:
```diff
@@ -0,0 +1,108 @@
+    }
+
+    private static ByteBuf toBase64Java8(ByteBuf src) {
+        ByteBuffer srcBuffer = src.internalNioBuffer(src.readerIndex(), src.readableBytes());
```
we are benchmarking `internalNioBuffer` but we are using `nioBuffer` in real code ... which is correct?

normanmaurer replied:
I think we should use internalBuffer always. Let me change.

Scottmitch commented:
> Maybe Oracle finally got something right ;)

We are currently optimizing for different scenarios. The JDK assumes all heap buffers, and we do some things which penalize heap buffers. For example we made some "optimizations" to avoid additional arithmetic/bounds checking by using the `setInt` type methods for direct buffers which is great, but for heap buffers this is actually more work (we must break apart the value into bytes and do the per byte indexing anyways). Starting from direct buffer and ensuring we end up with a direct buffer gives the following:

```
Benchmark                                          (level)  (size)   Mode  Cnt         Score        Error  Units
Base64Benchmark.javaBase64Encode                  DISABLED   65536  thrpt   10   8885823.621 ± 101862.732  ops/s
Base64Benchmark.nettyBase64EncodePooledDirectDst  DISABLED   65536  thrpt   10  11682215.993 ± 154439.862  ops/s
```

We should consider restoring the original read byte-by-byte code when reading from or writing to a heap buffer.

normanmaurer commented:
@Scottmitch I adjusted the benchmark and will re-run. I am not sure we should worry about heap buffers imho... Let us see how the numbers of the adjusted benchmark are.

normanmaurer commented:
ok benchmark started... will report back in a while :)

Scottmitch commented:
> I am not sure we should worry about heap buffers imho

It looks like certificates/keys are accepted only as encoded `byte[]` currently ... so maybe useful for the read side?

normanmaurer commented:
@Scottmitch sure it is useful ... just not think it is the most important for now. Let us see how the current benchmark does first.

normanmaurer commented:
New benchmark results (the adjusted benchmark):

```
# Run complete. Total time: 00:33:02

Benchmark                                (directSrc)   (level)  (size)   Mode  Cnt         Score        Error  Units
Base64Benchmark.javaBase64Encode               false  DISABLED    1024  thrpt   40   6371381.415 ?  32147.978  ops/s
Base64Benchmark.javaBase64Encode               false  DISABLED   65536  thrpt   40   6375275.156 ?  34154.503  ops/s
Base64Benchmark.javaBase64Encode               false    SIMPLE    1024  thrpt   40   6139242.983 ?  31225.194  ops/s
Base64Benchmark.javaBase64Encode               false    SIMPLE   65536  thrpt   40   6004022.069 ? 134762.735  ops/s
Base64Benchmark.javaBase64Encode               false  ADVANCED    1024  thrpt   40   1442311.580 ?  25429.857  ops/s
Base64Benchmark.javaBase64Encode               false  ADVANCED   65536  thrpt   40   1429420.613 ?  13796.172  ops/s
Base64Benchmark.javaBase64Encode               false  PARANOID    1024  thrpt   40     14978.056 ?    251.937  ops/s
Base64Benchmark.javaBase64Encode               false  PARANOID   65536  thrpt   40     14864.970 ?    215.804  ops/s
Base64Benchmark.javaBase64EncodePooled         false  DISABLED    1024  thrpt   40   7745195.117 ?  72997.374  ops/s
Base64Benchmark.javaBase64EncodePooled         false  DISABLED   65536  thrpt   40   8209046.167 ?  77797.495  ops/s
Base64Benchmark.javaBase64EncodePooled         false    SIMPLE    1024  thrpt   40   7749911.073 ?  85097.241  ops/s
Base64Benchmark.javaBase64EncodePooled         false    SIMPLE   65536  thrpt   40   7111614.963 ?  92159.423  ops/s
Base64Benchmark.javaBase64EncodePooled         false  ADVANCED    1024  thrpt   40   1659981.454 ?  24704.033  ops/s
Base64Benchmark.javaBase64EncodePooled         false  ADVANCED   65536  thrpt   40   1638345.343 ?  16202.938  ops/s
Base64Benchmark.javaBase64EncodePooled         false  PARANOID    1024  thrpt   40     15998.772 ?    180.456  ops/s
Base64Benchmark.javaBase64EncodePooled         false  PARANOID   65536  thrpt   40     15973.356 ?    188.775  ops/s
Base64Benchmark.nettyBase64Encode              false  DISABLED    1024  thrpt   40   7419112.983 ?  67211.852  ops/s
Base64Benchmark.nettyBase64Encode              false  DISABLED   65536  thrpt   40   7483553.011 ?  77235.531  ops/s
Base64Benchmark.nettyBase64Encode              false    SIMPLE    1024  thrpt   40   7197936.584 ?  51935.597  ops/s
Base64Benchmark.nettyBase64Encode              false    SIMPLE   65536  thrpt   40   7255528.575 ?  53151.662  ops/s
Base64Benchmark.nettyBase64Encode              false  ADVANCED    1024  thrpt   40   1251677.142 ?  13838.881  ops/s
Base64Benchmark.nettyBase64Encode              false  ADVANCED   65536  thrpt   40   1259908.800 ?  15427.366  ops/s
Base64Benchmark.nettyBase64Encode              false  PARANOID    1024  thrpt   40     11885.196 ?    140.708  ops/s
Base64Benchmark.nettyBase64Encode              false  PARANOID   65536  thrpt   40     11849.318 ?    164.850  ops/s
Base64Benchmark.nettyBase64EncodePooled        false  DISABLED    1024  thrpt   40  12956910.779 ? 185741.896  ops/s
Base64Benchmark.nettyBase64EncodePooled        false  DISABLED   65536  thrpt   40  14075229.771 ? 838265.617  ops/s
Base64Benchmark.nettyBase64EncodePooled        false    SIMPLE    1024  thrpt   40  12637737.290 ? 544555.683  ops/s
Base64Benchmark.nettyBase64EncodePooled        false    SIMPLE   65536  thrpt   40  12411013.463 ? 175626.718  ops/s
Base64Benchmark.nettyBase64EncodePooled        false  ADVANCED    1024  thrpt   40   1339411.523 ?  17779.528  ops/s
Base64Benchmark.nettyBase64EncodePooled        false  ADVANCED   65536  thrpt   40   1338455.468 ?  20721.495  ops/s
Base64Benchmark.nettyBase64EncodePooled        false  PARANOID    1024  thrpt   40     11768.081 ?    162.160  ops/s
Base64Benchmark.nettyBase64EncodePooled        false  PARANOID   65536  thrpt   40     11769.202 ?    180.615  ops/s
```

Scottmitch commented:
@normanmaurer - this may be a regression when using direct/pooled IIUC? as mentioned in https://github.com/netty/netty/pull/7030#issuecomment-320300566 we may also want to consider a followup in Netty's Base64 when reading from heap buffers.

normanmaurer commented:
alright... sorry it took a bit longer, but I updated some code to make sure it is also "fast" with paranoid leak detection and fixed the benchmark. Here are the latest results:

```

Result "io.netty.handler.codec.Base64Benchmark.nettyBase64EncodePooled":
  1.456 ±(99.9%) 0.031 ops/s [Average]
  (min, avg, max) = (1.342, 1.456, 1.534), stdev = 0.055
  CI (99.9%): [1.426, 1.487] (assumes normal distribution)


# Run complete. Total time: 00:52:24

Benchmark                                (directSrc)   (level)  (size)   Mode  Cnt       Score      Error  Units
Base64Benchmark.javaBase64Encode                true    SIMPLE    1024  thrpt   40  418261.443 ± 6217.186  ops/s
Base64Benchmark.javaBase64Encode                true    SIMPLE   65536  thrpt   40    9310.437 ±   83.420  ops/s
Base64Benchmark.javaBase64Encode                true  ADVANCED    1024  thrpt   40  327586.422 ± 4088.211  ops/s
Base64Benchmark.javaBase64Encode                true  ADVANCED   65536  thrpt   40    9170.043 ±   79.807  ops/s
Base64Benchmark.javaBase64Encode                true  PARANOID    1024  thrpt   40    6959.041 ±   75.516  ops/s
Base64Benchmark.javaBase64Encode                true  PARANOID   65536  thrpt   40    3896.663 ±   73.864  ops/s
Base64Benchmark.javaBase64Encode               false    SIMPLE    1024  thrpt   40  471896.485 ± 3867.918  ops/s
Base64Benchmark.javaBase64Encode               false    SIMPLE   65536  thrpt   40   10504.621 ±   77.521  ops/s
Base64Benchmark.javaBase64Encode               false  ADVANCED    1024  thrpt   40  367871.845 ± 2761.795  ops/s
Base64Benchmark.javaBase64Encode               false  ADVANCED   65536  thrpt   40   10469.282 ±  153.421  ops/s
Base64Benchmark.javaBase64Encode               false  PARANOID    1024  thrpt   40   12093.646 ±  139.551  ops/s
Base64Benchmark.javaBase64Encode               false  PARANOID   65536  thrpt   40    5562.653 ±   38.213  ops/s
Base64Benchmark.javaBase64EncodePooled          true    SIMPLE    1024  thrpt   40  480609.975 ±  998.443  ops/s
Base64Benchmark.javaBase64EncodePooled          true    SIMPLE   65536  thrpt   40    9738.197 ±   17.306  ops/s
Base64Benchmark.javaBase64EncodePooled          true  ADVANCED    1024  thrpt   40  359959.205 ± 4364.012  ops/s
Base64Benchmark.javaBase64EncodePooled          true  ADVANCED   65536  thrpt   40    9638.270 ±   32.304  ops/s
Base64Benchmark.javaBase64EncodePooled          true  PARANOID    1024  thrpt   40    7434.862 ±   92.519  ops/s
Base64Benchmark.javaBase64EncodePooled          true  PARANOID   65536  thrpt   40    4096.822 ±   32.269  ops/s
Base64Benchmark.javaBase64EncodePooled         false    SIMPLE    1024  thrpt   40  527980.032 ± 1209.189  ops/s
Base64Benchmark.javaBase64EncodePooled         false    SIMPLE   65536  thrpt   40   10701.257 ±   13.334  ops/s
Base64Benchmark.javaBase64EncodePooled         false  ADVANCED    1024  thrpt   40  391170.404 ± 2355.129  ops/s
Base64Benchmark.javaBase64EncodePooled         false  ADVANCED   65536  thrpt   40   10594.732 ±   21.420  ops/s
Base64Benchmark.javaBase64EncodePooled         false  PARANOID    1024  thrpt   40   12077.544 ±  164.403  ops/s
Base64Benchmark.javaBase64EncodePooled         false  PARANOID   65536  thrpt   40    5502.251 ±   87.693  ops/s
Base64Benchmark.nettyBase64Encode               true    SIMPLE    1024  thrpt   40  249026.275 ± 4153.030  ops/s
Base64Benchmark.nettyBase64Encode               true    SIMPLE   65536  thrpt   40    4854.955 ±  101.224  ops/s
Base64Benchmark.nettyBase64Encode               true  ADVANCED    1024  thrpt   40   11454.019 ±  793.965  ops/s
Base64Benchmark.nettyBase64Encode               true  ADVANCED   65536  thrpt   40     151.500 ±   46.428  ops/s
Base64Benchmark.nettyBase64Encode               true  PARANOID    1024  thrpt   40      49.374 ±    0.822  ops/s
Base64Benchmark.nettyBase64Encode               true  PARANOID   65536  thrpt   40       0.784 ±    0.017  ops/s
Base64Benchmark.nettyBase64Encode              false    SIMPLE    1024  thrpt   40  244054.056 ± 1501.413  ops/s
Base64Benchmark.nettyBase64Encode              false    SIMPLE   65536  thrpt   40    4276.909 ±  450.630  ops/s
Base64Benchmark.nettyBase64Encode              false  ADVANCED    1024  thrpt   40   11759.206 ±  661.290  ops/s
Base64Benchmark.nettyBase64Encode              false  ADVANCED   65536  thrpt   40     195.304 ±   61.906  ops/s
Base64Benchmark.nettyBase64Encode              false  PARANOID    1024  thrpt   40      93.533 ±    1.608  ops/s
Base64Benchmark.nettyBase64Encode              false  PARANOID   65536  thrpt   40       1.512 ±    0.022  ops/s
Base64Benchmark.nettyBase64EncodePooled         true    SIMPLE    1024  thrpt   40  247884.880 ± 7925.869  ops/s
Base64Benchmark.nettyBase64EncodePooled         true    SIMPLE   65536  thrpt   40    4616.999 ±   50.255  ops/s
Base64Benchmark.nettyBase64EncodePooled         true  ADVANCED    1024  thrpt   40   11129.730 ±  806.274  ops/s
Base64Benchmark.nettyBase64EncodePooled         true  ADVANCED   65536  thrpt   40     217.034 ±   86.414  ops/s
Base64Benchmark.nettyBase64EncodePooled         true  PARANOID    1024  thrpt   40      49.621 ±    0.610  ops/s
Base64Benchmark.nettyBase64EncodePooled         true  PARANOID   65536  thrpt   40       0.783 ±    0.009  ops/s
Base64Benchmark.nettyBase64EncodePooled        false    SIMPLE    1024  thrpt   40  267883.311 ±  986.460  ops/s
Base64Benchmark.nettyBase64EncodePooled        false    SIMPLE   65536  thrpt   40    4142.894 ±  449.079  ops/s
Base64Benchmark.nettyBase64EncodePooled        false  ADVANCED    1024  thrpt   40   11650.638 ±  756.135  ops/s
Base64Benchmark.nettyBase64EncodePooled        false  ADVANCED   65536  thrpt   40     195.972 ±   75.358  ops/s
Base64Benchmark.nettyBase64EncodePooled        false  PARANOID    1024  thrpt   40      91.517 ±    1.894  ops/s
Base64Benchmark.nettyBase64EncodePooled        false  PARANOID   65536  thrpt   40       1.456 ±    0.031  ops/s
```

So I think with this we have no need to use the JDK impl.

@johnou @Scottmitch @johnou @doom369 WDYT ?

normanmaurer commented:
sorry... need to re-run the benchmark. missed some change :(

johnou commented on pom.xml:
```diff
@@ -695,6 +695,8 @@
 
             <ignore>java.util.concurrent.ConcurrentLinkedDeque</ignore>
             <ignore>java.util.concurrent.ThreadLocalRandom</ignore>
+            <ignore>java.util.Base64</ignore>
```
Revert?

johnou commented:
@normanmaurer this good to pull in?

normanmaurer commented:
Not yet as I had no time to investigate... that said we did improve the perf of the leak detector so I think in general we really fixed the issue mentioned here

> Am 07.10.2017 um 18:33 schrieb Johno Crawford <notifications@github.com>:
> 
> @normanmaurer this good to pull in?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

normanmaurer commented:
Let me close this for now as the "main reason" for this PR is fixed as we improved the speed of the leak-detector.
